src/IO/LzmaReadBuffer.cpp

```diff
@@ -52,16 +52,21 @@ bool LzmaReadBuffer::nextImpl()
     lstr.next_out = reinterpret_cast<unsigned char *>(internal_buffer.begin());
     lstr.avail_out = internal_buffer.size();
 
-    lzma_ret ret = lzma_code(&lstr, LZMA_FINISH);
+    lzma_ret ret = lzma_code(&lstr, LZMA_RUN);
 
     in->position() = in->buffer().end() - lstr.avail_in;
+    working_buffer.resize(internal_buffer.size() - lstr.avail_out);
 
     if (ret == LZMA_STREAM_END)
     {
         if (in->eof())
         {
             eof = true;
             return working_buffer.size() != 0;
+        } else {
+            throw Exception(
+                std::string("lzma decoder finished, but stream is still alive: error code: ") + std::to_string(ret) + "; lzma version: " + LZMA_VERSION_STRING,
+                ErrorCodes::LZMA_STREAM_DECODER_FAILED);
```

Comment:
This can use fmt-like constructor

```suggestion
            throw Exception(ErrorCodes::LZMA_STREAM_DECODER_FAILED, "lzma decoder finished, but stream is still alive: error code: {}; lzma version: {}", ret, LZMA_VERSION_STRING);
```